File: Dev/unitTest_Engine_cs/ObjectSystem2D/LifeCycle/DisposeOnTerminate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using asd;

namespace unitTest_Engine_cs.ObjectSystem2D.LifeCycle
{
	class DisposeOnTerminate : TestFramework
	{
		class TestScene : Scene
		{
			public bool AlreadyDisposed = false;

			protected override void OnDispose()
			{
				AlreadyDisposed = true;
			}
		}

		public override void Test(GraphicsDeviceType graphicsType)
		{
			var option = new EngineOption()
			{
				GraphicsDevice = graphicsType
			};
			Engine.Initialize("DisposeOnTerminate", 640, 480, option);

			var scene = new TestScene();
			Engine.ChangeScene(scene);

			while (Engine.DoEvents())
			{
				Engine.Update();
			}

			Engine.Terminate();

			Assert.AreEqual(true, scene.AlreadyDisposed);
		}
	}
}

File: Dev/unitTest_Engine_cpp_gtest/ObjectSystem/DisposeOnTerminate.cpp
#include <gtest/gtest.h>
#include <Altseed.h>
#include "../EngineTest.h"

using namespace std;
using namespace asd;

class DisposeOnTerminate_Scene : public Scene
{
public:
	bool m_alreadyDisposed = false;

protected:
	void OnDispose() override
	{
		m_alreadyDisposed = true;
	}
};

void Test_ObjectSystem_DisposeOnTerminate(bool openGL)
{
	auto option = EngineOption();
	option.GraphicsDevice = openGL ? GraphicsDeviceType::OpenGL : GraphicsDeviceType::DirectX11;
	Engine::Initialize(ToAString("DisposeOnTerminate").c_str(), 640, 480, option);

	auto scene = make_shared<DisposeOnTerminate_Scene>();
	Engine::ChangeScene(scene);

	while (Engine::DoEvents())
	{
		Engine::Update();
	}

	Engine::Terminate();

	ASSERT_EQ(true, scene->m_alreadyDisposed);
}